### PR TITLE
Add integration test for OIDC id token claim validation when session  is available

### PR DESCRIPTION
**Pull request by GDRDABARERA**
Merge this test-case, after the issue https://github.com/wso2/product-is/issues/3058 is fixed

### CODE REVIEW

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/oidc/OIDCAuthzCodeIdTokenValidationTestCase.java [hunk 1]
```diff
@@ -71,12 +71,12 @@
...
-    private static final String CALLBACK_URL = "https://localhost/callback";
-    private CloseableHttpClient client;
-    private String sessionDataKey;
-    private String sessionDataKeyConsent;
-    private AuthorizationCode authorizationCode;
-    private String idToken;
+    protected static final String CALLBACK_URL = "https://localhost/callback";
+    protected CloseableHttpClient client;
+    protected String sessionDataKey;
+    protected  String sessionDataKeyConsent;
```
Comment (pulasthi7) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/oidc/OIDCAuthzCodeIdTokenValidationTestCase.java]: fix formatting

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/oidc/OIDCAuthzCodeIdTokenValidationTestCase.java [hunk 2]
```diff
@@ -71,12 +71,12 @@
     public static final String TEST_NONCE = "test_nonce";
     private RSAPrivateKey spPrivateKey;
     private X509Certificate spX509PublicCert;
-    private static final String CALLBACK_URL = "https://localhost/callback";
-    private CloseableHttpClient client;
-    private String sessionDataKey;
-    private String sessionDataKeyConsent;
-    private AuthorizationCode authorizationCode;
-    private String idToken;
+    protected static final String CALLBACK_URL = "https://localhost/callback";
```
Comment (pulasthi7) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/oidc/OIDCAuthzCodeIdTokenValidationTestCase.java]: move constant definitions before class variables

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/oidc/OIDCAuthzCodeIdTokenValidationWhenSessionExistsTestCase.java [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+import org.apache.http.message.BasicNameValuePair;
+import org.testng.Assert;
+import org.testng.annotations.Test;
+import org.wso2.identity.integration.test.utils.DataExtractUtil;
+import org.wso2.identity.integration.test.utils.OAuth2Constant;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class OIDCAuthzCodeIdTokenValidationWhenSessionExistsTestCase extends OIDCAuthzCodeIdTokenValidationTestCase {
```
Comment (pulasthi7) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/oidc/OIDCAuthzCodeIdTokenValidationWhenSessionExistsTestCase.java]: Extending an existing test will run its test again in this class also. Get the common methods to an abstract class and extend both these classes from that abstract class

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/oidc/OIDCAuthzCodeIdTokenValidationWhenSessionExistsTestCase.java [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+
+    @Test(groups = "wso2.is", description = "Send authorize user request for authorization code grant type.")
+    public void testNewAuthCodeGrantSendAuthRequestPost() throws Exception {
+
+        testSecondAuthCodeGrantSendAuthRequestPost();
+        testAuthCodeGrantSendApprovalPost();
+        testAuthCodeGrantSendGetTokensPost();
```
Comment (pulasthi7) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/oidc/OIDCAuthzCodeIdTokenValidationWhenSessionExistsTestCase.java]: shouldn't these methods be individual tests?

**Comment by pulasthi7** [comment]
Following tests are failing when this PR is merged locally, please have a look and fix.

Tests run: 569, Failures: 3, Errors: 0, Skipped: 0, Time elapsed: 7,045.454 sec <<< FAILURE!
testRevokeTokenAfterCacheTimedOut(org.wso2.identity.integration.test.oauth2.OAuth2TokenRevokeAfterCacheTimeOutTestCase)  Time elapsed: 0.481 sec  <<< FAILURE!
java.lang.Exception: IdentityApplicationManagementServiceIdentityApplicationManagementException
	at org.wso2.identity.integration.common.clients.application.mgt.ApplicationManagementServiceClient.createApplication(ApplicationManagementServiceClient.java:126)
	at org.wso2.identity.integration.test.oauth2.OAuth2ServiceAbstractIntegrationTest.createApplication(OAuth2ServiceAbstractIntegrationTest.java:149)
	at org.wso2.identity.integration.test.oauth2.OAuth2ServiceAbstractIntegrationTest.createApplication(OAuth2ServiceAbstractIntegrationTest.java:124)
	at org.wso2.identity.integration.test.oauth2.OAuth2TokenRevokeAfterCacheTimeOutTestCase.testRevokeTokenAfterCacheTimedOut(OAuth2TokenRevokeAfterCacheTimeOutTestCase.java:67)

testNewAuthCodeGrantSendAuthRequestPost(org.wso2.identity.integration.test.oidc.OIDCAuthzCodeIdTokenValidationWhenSessionExistsTestCase)  Time elapsed: 0.08 sec  <<< FAILURE!
java.lang.AssertionError: expected object to not be null
	at org.testng.Assert.fail(Assert.java:89)
	at org.testng.Assert.assertNotNull(Assert.java:399)
	at org.testng.Assert.assertNotNull(Assert.java:384)
	at org.wso2.identity.integration.test.oidc.OIDCAuthzCodeIdTokenValidationTestCase.getLocationHeaderValue(OIDCAuthzCodeIdTokenValidationTestCase.java:243)
	at org.wso2.identity.integration.test.oidc.OIDCAuthzCodeIdTokenValidationWhenSessionExistsTestCase.testSecondAuthCodeGrantSendAuthRequestPost(OIDCAuthzCodeIdTokenValidationWhenSessionExistsTestCase.java:63)
	at org.wso2.identity.integration.test.oidc.OIDCAuthzCodeIdTokenValidationWhenSessionExistsTestCase.testNewAuthCodeGrantSendAuthRequestPost(OIDCAuthzCodeIdTokenValidationWhenSessionExistsTestCase.java:39)

atEnd(org.wso2.identity.integration.test.oidc.OIDCAuthzCodeIdTokenValidationWhenSessionExistsTestCase)  Time elapsed: 0.099 sec  <<< FAILURE!
java.lang.Exception: Transport error: 401 Error: Unauthorized
	at org.wso2.identity.integration.common.clients.application.mgt.ApplicationManagementServiceClient.deleteApplication(ApplicationManagementServiceClient.java:194)
	at org.wso2.identity.integration.test.oauth2.OAuth2ServiceAbstractIntegrationTest.deleteApplication(OAuth2ServiceAbstractIntegrationTest.java:437)
	at org.wso2.identity.integration.test.oidc.OIDCAuthzCodeIdTokenValidationTestCase.atEnd(OIDCAuthzCodeIdTokenValidationTestCase.java:90)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:80)
	at org.testng.internal.Invoker.invokeConfigurationMethod(Invoker.java:525)
	at org.testng.internal.Invoker.invokeConfigurations(Invoker.java:202)
	at org.testng.internal.Invoker.invokeConfigurations(Invoker.java:130)
	at org.testng.internal.TestMethodWorker.invokeAfterClassMethods(TestMethodWorker.java:222)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:112)
	at org.testng.TestRunner.runWorkers(TestRunner.java:1178)
	at org.testng.TestRunner.privateRun(TestRunner.java:757)
	at org.testng.TestRunner.run(TestRunner.java:608)
	at org.testng.SuiteRunner.runTest(SuiteRunner.java:334)
	at org.testng.SuiteRunner.runSequentially(SuiteRunner.java:329)
	at org.testng.SuiteRunner.privateRun(SuiteRunner.java:291)
	at org.testng.SuiteRunner.run(SuiteRunner.java:240)
	at org.testng.SuiteRunnerWorker.runSuite(SuiteRunnerWorker.java:52)
	at org.testng.SuiteRunnerWorker.run(SuiteRunnerWorker.java:86)
	at org.testng.TestNG.runSuitesSequentially(TestNG.java:1158)
	at org.testng.TestNG.runSuitesLocally(TestNG.java:1083)
	at org.testng.TestNG.run(TestNG.java:999)
	at org.apache.maven.surefire.testng.TestNGExecutor.run(TestNGExecutor.java:178)
	at org.apache.maven.surefire.testng.TestNGXmlTestSuite.execute(TestNGXmlTestSuite.java:92)
	at org.apache.maven.surefire.testng.TestNGProvider.invoke(TestNGProvider.java:96)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.apache.maven.surefire.util.ReflectionUtils.invokeMethodWithArray(ReflectionUtils.java:189)
	at org.apache.maven.surefire.booter.ProviderFactory$ProviderProxy.invoke(ProviderFactory.java:165)
	at org.apache.maven.surefire.booter.ProviderFactory.invokeProvider(ProviderFactory.java:85)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:115)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:75)


Results :

Failed tests:   testRevokeTokenAfterCacheTimedOut(org.wso2.identity.integration.test.oauth2.OAuth2TokenRevokeAfterCacheTimeOutTestCase): IdentityApplicationManagementServiceIdentityApplicationManagementException
  testNewAuthCodeGrantSendAuthRequestPost(org.wso2.identity.integration.test.oidc.OIDCAuthzCodeIdTokenValidationWhenSessionExistsTestCase): expected object to not be null
  atEnd(org.wso2.identity.integration.test.oidc.OIDCAuthzCodeIdTokenValidationWhenSessionExistsTestCase): Transport error: 401 Error: Unauthorized

Tests run: 569, Failures: 3, Errors: 0, Skipped: 0

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/oidc/OIDCAuthzCodeIdTokenValidationWhenSessionExistsTestCase.java (hunk 3 above)
Comment (GDRDABARERA) [reply modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/oidc/OIDCAuthzCodeIdTokenValidationWhenSessionExistsTestCase.java]: The main objective in here is to run the supper class once and again run the same methods in the same session. So creating an abstract class won't achieve that objective.

**Comment by GDRDABARERA** [comment]
Yes when we merge this test it fails because there is an issue in our product. We need to fix that first before merging this https://github.com/wso2/product-is/pull/3055

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/oidc/OIDCAuthzCodeIdTokenValidationWhenSessionExistsTestCase.java (hunk 4 above)
Comment (GDRDABARERA) [reply modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/oidc/OIDCAuthzCodeIdTokenValidationWhenSessionExistsTestCase.java]: They themselves are test-cases mentioned in supper class,  So is it intercessory to make it individual test cases again?

**Comment by GDRDABARERA** [comment]
Closing this till the bug fixes

**Pull request closed without merging**